### PR TITLE
SOLR-14404 CoreContainer level custom requesthandlers

### PR DESCRIPTION
**Pull request by noblepaul**

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/api/ApiBag.java [hunk 1]
```diff
@@ -134,6 +142,14 @@ static void registerIntrospect(List<String> l, PathTrie<Api> registry, Map<Strin
     registry.insert(copy, substitutes, introspect);
   }
 
+  public Api unregister(SolrRequest.METHOD method, String path) {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/api/ApiBag.java:145]: Should this be synchronized to match `register`?

File: solr/core/src/java/org/apache/solr/handler/admin/ContainerPluginsApi.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+          payload.addError(info.name + " already exists");
+          return null;
+        }
+        map.put(info.name, info);
+        return map;
+      });
+    }
+
+    @Command(name = "remove")
+    public void remove(SolrQueryRequest req, SolrQueryResponse rsp, PayloadObj<String> payload) throws IOException {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/handler/admin/ContainerPluginsApi.java:97]: Should this be METHOD.DELETE instead of a post?

File: solr/core/src/java/org/apache/solr/pkg/PackageListeners.java [hunk 3]
```diff
@@ -63,13 +63,13 @@ public synchronized void removeListener(Listener listener) {
   }
 
   synchronized void packagesUpdated(List<PackageLoader.Package> pkgs) {
-    MDCLoggingContext.setCore(core);
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/pkg/PackageListeners.java:66]: Why do we need this check? setCore already handles nulls.

File: solr/core/src/test-files/runtimecode/sig.txt [hunk 4]
```diff
@@ -69,6 +69,14 @@ openssl dgst -sha1 -sign ../cryptokeys/priv_key512.pem expressible.jar.bin | ope
 
 ZOT11arAiPmPZYOHzqodiNnxO9pRyRozWZEBX8XGjU1/HJptFnZK+DI7eXnUtbNaMcbXE2Ze8hh4M/eGyhY8BQ==
 
+openssl dgst -sha1 -sign priv_key512.pem containerplugin.v.1.jar.bin | openssl enc -base64 | sed 's/+/%2B/g' | tr -d \\n | sed
```
Comment (madrob) [inline solr/core/src/test-files/runtimecode/sig.txt:72]: Does this need to be `../cryptokeys/priv_key512.pem`?

File: solr/core/src/test-files/runtimecode/MyPlugin.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import org.apache.solr.client.solrj.SolrRequest.METHOD;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.apache.solr.security.PermissionNameProvider;
+
+@EndPoint(path = "/plugin/my/path",
+    method = METHOD.GET,
+    permission = PermissionNameProvider.Name.CONFIG_READ_PERM)
+public class MyPlugin {
```
Comment (madrob) [inline solr/core/src/test-files/runtimecode/MyPlugin.java:31]: I assume this is used to generate containerplugin.v.1.jar.bin and v.2? Should we have two source files for those? Can we do this some other way, besides checking in binaries to the repo?

File: solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    @Command
+    public void read(SolrQueryRequest req, SolrQueryResponse rsp) {
+      rsp.add("testkey", "testval");
+    }
+
+  }
+
+
+  private void expectError(V2Request req, SolrClient client, String errPath, String expectErrorMsg) throws IOException, SolrServerException {
+    try {
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java:242]: use `expectThrows` here.

File: solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+  }
+
+  @After
+  public void teardown() {
+    System.clearProperty("enable.packages");
+  }
+
+  @Test
+  public void testApi() throws Exception {
+    MiniSolrCloudCluster cluster =
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java:69]: cluster setup and teardown can go in Before/After methods

File: solr/core/src/test/org/apache/solr/handler/admin/TestApiFramework.java [hunk 8]
```diff
@@ -199,6 +200,25 @@ public void testPayload() {
 
   }
 
+  public void testApiWrapper() {
+    Class<ApiWithConstructor> klas = ApiWithConstructor.class;
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/handler/admin/TestApiFramework.java:204]: unclear what this test is testing

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/BaseHttpSolrClient.java [hunk 9]
```diff
@@ -62,6 +63,9 @@ public static RemoteExecutionException create(String host, NamedList errResponse
       if (errObj != null) {
         Number code = (Number) getObjectByPath(errObj, true, Collections.singletonList("code"));
         String msg = (String) getObjectByPath(errObj, true, Collections.singletonList("msg"));
+        if(msg == null) msg = "";
```
Comment (madrob) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/BaseHttpSolrClient.java:66]: There's already a null check later, what kinds of messages do we get here that are useful and don't leak too much internals?

File: solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+          .build()
+          .process(cluster.getSolrClient());
+      assertEquals(null, rsp._get("/plugin/testplugin/class", null));
+
+    } finally {
+      cluster.shutdown();
+    }
+  }
+  @Test
+  public void testApiFromPackage() throws Exception {
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java:124]: this is a fairly long test and was somewhat hard to follow, can we break it up into smaller pieces?

File: solr/core/src/java/org/apache/solr/core/CoreContainer.java [hunk 11]
```diff
@@ -648,8 +655,8 @@ public void load() {
       pkiAuthenticationPlugin.initializeMetrics(solrMetricsContext, "/authentication/pki");
       TracerConfigurator.loadTracer(loader, cfg.getTracerConfiguratorPluginInfo(), getZkController().getZkStateReader());
       packageLoader = new PackageLoader(this);
-      containerHandlers.getApiBag().register(new AnnotatedApi(packageLoader.getPackageAPI().editAPI), Collections.EMPTY_MAP);
-      containerHandlers.getApiBag().register(new AnnotatedApi(packageLoader.getPackageAPI().readAPI), Collections.EMPTY_MAP);
+      containerHandlers.getApiBag().registerObject(packageLoader.getPackageAPI().editAPI);
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/core/CoreContainer.java:658]: There's a minor inconsistency here between edit/read and write/read, maybe we can standardize on one?

File: solr/core/src/java/org/apache/solr/handler/admin/ContainerPluginsApi.java (hunk 2 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/handler/admin/ContainerPluginsApi.java:97]: We could, but , this is the convention that we follow

File: solr/core/src/java/org/apache/solr/pkg/PackageListeners.java (hunk 3 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/pkg/PackageListeners.java:66]: because we are calling a 

`MDCLoggingContext.clear();` in finally

File: solr/core/src/test-files/runtimecode/MyPlugin.java (hunk 5 above)
Comment (noblepaul) [reply solr/core/src/test-files/runtimecode/MyPlugin.java:31]: This is for documentation purposes for anyone who wants to know what the code looks like. Yes, we could possibly have 2 files, but then , they will have to be different files

> Can we do this some other way,

Unfortunately no. We need to have it compiled with an old version of java. It has to be precreated

>

File: solr/core/src/java/org/apache/solr/pkg/PackageListeners.java (hunk 3 above)
Comment (madrob) [reply solr/core/src/java/org/apache/solr/pkg/PackageListeners.java:66]: I was under the impression that it setCore/clear still work even if the core was null.

File: solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+
+      String path = pluginNameVsPath.get(e.getKey());
+      if (path == null) {
+        // there is a new apiInfo . let's register it
+        try {
+          apiInfo.init();
+          ApiHolder holder = new ApiHolder(apiInfo);
+          plugins.put(holder.key, holder);
+          pluginNameVsPath.put(apiInfo.info.name, holder.key);
+          containerApiBag.register(holder, Collections.EMPTY_MAP);
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java:117]: prefer Collections.emptyMap()

File: solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+  public void writeMap(EntryWriter ew) {
+    plugins.forEach((s, apiHolder) -> ew.putNoEx(s, apiHolder.apiInfo));
+  }
+
+  public CustomContainerPlugins(CoreContainer coreContainer, ApiBag apiBag) {
+    this.coreContainer = coreContainer;
+    this.containerApiBag = apiBag;
+  }
+
+  public void refresh(Map<String, Object> pluginInfos) {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java:79]: This parameter is unused?

File: solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+    private PackageLoader.Package.Version pkgVersion;
+    EndPoint endPoint;
+
+    private Class klas;
+
+
+    private AnnotatedApi delegate;
+
+
+    public ApiInfo(PluginMeta info, List<String> errs) {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java:201]: This feels like a super klunky API writing errors to a string in a constructor argument? Why can't we just throw exceptions here?

File: solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+    public String key;
+    @JsonProperty
+    private final PluginMeta info;
+
+    @JsonProperty(value = "package")
+    public final String pkg;
+    private PackageLoader.Package.Version pkgVersion;
+    EndPoint endPoint;
+
+    private Class klas;
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java:195]: nit: raw types

File: solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+      return;
+    }
+    if(pluginInfos.isEmpty()) return;
+
+    for (Map.Entry<String, Object> e : pluginInfos.entrySet()) {
+      PluginMeta info = null;
+      try {
+        info = mapper.readValue(Utils.toJSON(e.getValue()), PluginMeta.class);
+      } catch (IOException ioException) {
+        log.error("Invalid apiInfo configuration :", ioException);
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java:93]: should this have a `continue`?

File: solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  final CoreContainer coreContainer;
+  final ApiBag containerApiBag;
+  private Map<String, ApiHolder> plugins = new HashMap<>();
+  private Map<String, String> pluginNameVsPath = new HashMap<>();
+
+  @Override
+  public boolean onChange(Map<String, Object> properties) {
+    refresh(null);
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java:65]: `refresh(properties)`?

File: solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+        }
+      }
+    }
+    Set<String> toBeRemoved = new HashSet<>();
+    for (String s : pluginNameVsPath.keySet()) {
+      if (!pluginInfos.containsKey(s)) {
+        toBeRemoved.add(s);
+      }
+    }
+    for (String s : toBeRemoved) {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java:147]: why do we need to iterate twice here?

File: solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+import org.apache.solr.pkg.PackageLoader;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.apache.solr.util.SolrJacksonAnnotationInspector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.lucene.util.IOUtils.closeWhileHandlingException;
+
+public class CustomContainerPlugins implements MapWriter, ClusterPropertiesListener {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java:54]: there's a lot of nested classes here, can any of those be split out into their own files?

File: solr/core/src/java/org/apache/solr/handler/admin/ContainerPluginsApi.java [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+      Map<String, Object> clusterPropsJson = (Map<String, Object>) Utils.fromJSON(zkClient.getData(ZkStateReader.CLUSTER_PROPS, null, new Stat(), true));
+      return (Map<String, Object>) clusterPropsJson.computeIfAbsent(PLUGIN, Utils.NEW_LINKED_HASHMAP_FUN);
+    } catch (KeeperException.NoNodeException e) {
+      return new LinkedHashMap<>();
+    } catch (KeeperException | InterruptedException e) {
+      throw new IOException("Error reading cluster property", SolrZkClient.checkInterrupted(e));
+    }
+  }
+
+  private void persistPlugins(Function<Map, Map> modifier) throws IOException {
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/handler/admin/ContainerPluginsApi.java:162]: nit: raw types on Map

File: solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java (hunk 14 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java:201]: The objective is to collect multiple errors and eventually throw one single Exception whihc contains all of them

File: solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java (hunk 7 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java:69]: These clusters are created on a per testcase basis. We don't want remnants of other tests to screw up other tests

File: solr/core/src/java/org/apache/solr/pkg/PackageListeners.java (hunk 3 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/pkg/PackageListeners.java:66]: OK, got it. I misunderstood the comment

File: solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java (hunk 10 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java:124]: Splitting these tests will create another cluster and slow down the test execution further. I shall try to add inline comments to make it more clear as to what it is doing

File: solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java (hunk 6 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java:242]: `expectThrows` does not work here

File: solr/core/src/test/org/apache/solr/handler/admin/TestApiFramework.java (hunk 8 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/handler/admin/TestApiFramework.java]: it got in there accidentally , removed

File: solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java (hunk 17 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java]: cleaned up

File: solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java (hunk 16 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/api/CustomContainerPlugins.java]: fixed

File: solr/core/src/java/org/apache/solr/pkg/PackageListeners.java [hunk 21]
```diff
@@ -69,7 +69,7 @@ synchronized void packagesUpdated(List<PackageLoader.Package> pkgs) {
         invokeListeners(pkgInfo);
       }
     } finally {
-      MDCLoggingContext.clear();
+      if(core != null) MDCLoggingContext.clear();
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/pkg/PackageListeners.java:72]: No; it was correct before.  Internally, setCore can accept null and doesn't require the caller to care if core is null.  Furthermore, PackageListeners.core will never be null.  It ought to be declared final as well; it's only set in the constructor.

File: solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java (hunk 6 above)
Comment (dsmiley) [reply solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java:300]: How so?  It returns the exception of the expected type and then you can inspect it as you are doing now.

File: solr/solrj/src/java/org/apache/solr/client/solrj/request/beans/PluginMeta.java [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+package org.apache.solr.client.solrj.request.beans;
+
+import java.util.Objects;
+
+import org.apache.solr.common.annotation.JsonProperty;
+import org.apache.solr.common.util.ReflectMapWriter;
+
+public class PluginMeta implements ReflectMapWriter {
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/client/solrj/request/beans/PluginMeta.java:25]: Add at least one line of javadoc for every public class

File: solr/solrj/src/java/org/apache/solr/common/util/PathTrie.java [hunk 23]
```diff
@@ -82,6 +80,22 @@ public boolean add(String s) {
     return parts;
   }
 
+  public T unregister(List<String> path) {
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/common/util/PathTrie.java:83]: I think the name "unregister" would only make sense if there was a corresponding "register" method.  How about a more generic "remove"?

**Comment by noblepaul** [comment]
thanks @dsmiley , I shall fix them

File: solr/core/src/java/org/apache/solr/pkg/PackageListeners.java (hunk 21 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/pkg/PackageListeners.java:72]: > PackageListeners.core will never be null

Now it's not.

The schema plugins will no more be associated with a core. They will have it as `null`. I don't need to deal with it in this. So, I'll remove the null check now. `PackageListeners` is not used in ContainerPlugins

File: solr/solrj/src/java/org/apache/solr/common/util/PathTrie.java (hunk 23 above)
Comment (noblepaul) [reply solr/solrj/src/java/org/apache/solr/common/util/PathTrie.java:83]: yeah, `remove()` is better for `insert()`

File: solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java (hunk 6 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/handler/TestContainerPlugin.java]: fixed

**Comment by noblepaul** [comment]
@dsmiley 
Yes, I want all our components to be written using

- Annotations (even our internal components)
- All input output to use POJOs
